### PR TITLE
Saliency fix, gradient error propagation fix, visualization, documentation and code style fix

### PR DESCRIPTION
**Pull request by rfratila**
- Removed double softmax transformations
  - Moved  to only `forward_pass` since criterion already applies it for training
- Removed default activation for `pred_activation` so we don't do Softmax(Relu(x)) during training. Left the variable for if it's an intermediate layer
- Added more tests

### CODE REVIEW

File: vulcanai2/models/basenetwork.py [hunk 1]
```diff
@@ -453,11 +454,16 @@ def forward_pass(self, input_data, convert_to_class=False):
         Returns: Numpy matrix with the output probabilities
                  with each class unless otherwise specified.
         """
-        output = self.cpu()(torch.Tensor(input_data)).data.numpy()
-        if convert_to_class:
-            return self.metrics.get_class(output)
-        else:
-            return output
+        self.eval()
+        if not isinstance(input_data, torch.Tensor):
```
Comment (nirtiac) [inline vulcanai2/models/basenetwork.py]: input data should be the same as fit, aka take a data_loader and perform necessary operations to get the input_data. Also add a parameter for number of samples to evaluate on.

File: vulcanai2/plotters/utils.py [hunk 2]
```diff
@@ -64,21 +71,28 @@ def remove_hooks(self):
         for h in self.hooks:
             h.remove()
 
-    def generate_gradients(self, input_x, target_class_list):
+    def generate_gradients(self, input_x, targets):
```
Comment (nirtiac) [inline vulcanai2/plotters/utils.py]: this should also take the same input type... and we should be consistent in our naming of variables and parameters in regards to data/x/training and targets/y data.

File: vulcanai2/models/cnn.py [hunk 3]
```diff
@@ -105,7 +105,9 @@ def _create_network(self, **kwargs):
 
     def _create_classification_layer(self, dim, pred_activation):
         self.network_tail = DenseUnit(
-            dim, self.out_dim, activation=pred_activation)
+            in_features=dim,
+            out_features=self.out_dim,
+            activation=pred_activation)
```
Comment (nirtiac) [inline vulcanai2/models/cnn.py]: TODO: check for correct dimensions between input data and self.dimensions. Also remove self.in_dim from _create_network

File: vulcanai2/models/metrics.py [hunk 4]
```diff
@@ -78,6 +80,8 @@ def get_class(self, in_matrix):
         :param n_matrix: one-hot matrix
         :return: 1D Class array
         """
+        if isinstance(in_matrix, torch.Tensor):
+            in_matrix = in_matrix.detach().numpy()
```
Comment (priyatharsan) [inline vulcanai2/models/metrics.py]: Getting an error at this point after commenting out `model1.runtest` in the example
```
Traceback (most recent call last):
  File "fashion_conv_dense_test.py", line 136, in <module>
    model1.forward_pass(val_loader, convert_to_class=True),
  File "c:\work\aifred\vulcan2\master\vulcan2\vulcanai2\models\basenetwork.py", line 473, in forward_pass
    self.metrics.get_class(in_matrix=predictions)).float()
  File "c:\work\aifred\vulcan2\master\vulcan2\vulcanai2\models\metrics.py", line 84, in get_class
    in_matrix = in_matrix.detach().numpy()
TypeError: can't convert CUDA tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.
```
Same CUDA issue.

File: vulcanai2/plotters/utils.py [hunk 5]
```diff
@@ -22,66 +23,81 @@ def __init__(self, network):
...
+        self.network.eval()
+        # To properly pass the gradients
+        if not isinstance(input_data, torch.Tensor):
+            input_data = torch.tensor(input_data, requires_grad=True)
+        else:
+            if not input_data.requires_grad:
+                input_data.requires_grad = True
         # Forward pass
-        network_output = self.network(input_x)
+        network_output = self.network(input_data)
```
Comment (priyatharsan) [inline vulcanai2/plotters/utils.py]: Change this to: `network_output = self.network.cpu()(input_data)`

File: vulcanai2/models/basenetwork.py [hunk 6]
```diff
@@ -434,30 +434,51 @@ def _validate(self, val_loader):
...
+            # Get raw network output
+            predictions = self(data)
+            if self._num_classes:
+                # Get probabilities
+                predictions = nn.Softmax(dim=1)(predictions)
+                if convert_to_class:
+                    predictions = torch.tensor(
+                        self.metrics.get_class(in_matrix=predictions)).float()
+            # Aggregate predictions
+            pred_collector = torch.cat([pred_collector, predictions])
```
Comment (priyatharsan) [inline vulcanai2/models/basenetwork.py]: Change to: `pred_collector = torch.cat([pred_collector, predictions.cpu()])`

File: vulcanai2/models/metrics.py (hunk 4 above)
Comment (priyatharsan) [reply vulcanai2/models/metrics.py]: Actually change it to: `in_matrix = in_matrix.cpu().detach().numpy()`

Comment (rfratila) [reply vulcanai2/models/metrics.py]: im just gonna change it before it gets called into get_class

File: vulcanai2/plotters/utils.py (hunk 5 above)
Comment (rfratila) [reply vulcanai2/plotters/utils.py]: will do

File: vulcanai2/models/basenetwork.py (hunk 6 above)
Comment (rfratila) [reply vulcanai2/models/basenetwork.py]: will do

**Pull request merged**